sdk/src/Core/Amazon.Runtime/Credentials/_bcl45+netstandard/SSOAWSCredentials.cs

```diff
@@ -0,0 +1,344 @@
+        /// <param name="roleName">The corresponding IAM role in the AWS account that temporary AWS credentials will be resolved for.</param>
+        /// <param name="startUrl">The main URL for users to login to their SSO directory.</param>
+        public SSOAWSCredentials(string accountId, string region, string roleName, string startUrl)
+            : this(accountId, region, roleName, startUrl, new SSOAWSCredentialsOptions())
+        {
+        }
+
+        /// <summary>
+        /// Constructs an SSOAWSCredentials object.
+        /// </summary>
+        /// <param name="accountId">The AWS account ID that temporary AWS credentials will be resolved for.</param>
+        /// <param name="region">The AWS region where the SSO directory for <paramref name="startUrl"/> is hosted.</param>
+        /// <param name="roleName">The corresponding IAM role in the AWS account that temporary AWS credentials will be resolved for.</param>
+        /// <param name="startUrl">The main URL for users to login to their SSO directory.</param>
+        /// <param name="options">Options to be used in the SSO flow to resolve credentials.</param>
+        public SSOAWSCredentials(
+            string accountId, string region,
+            string roleName, string startUrl,
+            SSOAWSCredentialsOptions options)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            AccountId = accountId;
+            Region = region;
+            RoleName = roleName;
+            StartUrl = startUrl;
+            Options = options;
+        }
+
+#if BCL
+        protected override CredentialsRefreshState GenerateNewCredentials()
+        {
+            ValidateCredentialsInputs();
+
+            var oidcClient = CreateClient<ICoreAmazonSSOOIDC>(
+                ServiceClientHelpers.SSO_OIDC_SERVICE_CLASS_NAME,
+                ServiceClientHelpers.SSO_OIDC_SERVICE_CONFIG_NAME,
+                ServiceClientHelpers.SSO_OIDC_ASSEMBLY_NAME);
+
+            var ssoClient = CreateClient<ICoreAmazonSSO>(
+                ServiceClientHelpers.SSO_SERVICE_CLASS_NAME,
+                ServiceClientHelpers.SSO_SERVICE_CONFIG_NAME,
+                ServiceClientHelpers.SSO_ASSEMBLY_NAME);
+
+            var credentials = GetSsoCredentials(oidcClient, ssoClient) as SSOImmutableCredentials;
+
+            if (credentials == null)
+            {
+                throw new NotSupportedException("Unable to get credentials from SSO");
+            }
+
+            _logger.InfoFormat("New SSO credentials created that expire at {0}", credentials.Expiration.ToString("yyyy-MM-ddTHH:mm:ss.fffffffK", CultureInfo.InvariantCulture));
+            return new CredentialsRefreshState(credentials, credentials.Expiration);
+        }
+#endif
+
+        protected override async Task<CredentialsRefreshState> GenerateNewCredentialsAsync()
```

Comment:
Do we need to do anything to prevent multiple SSO flows from happening at the same time for the same SSO options? We a version of this in the DefaultInstanceProfileAWSCredentials which uses a singleton and a timer to refresh credentials. Not saying we should do the same design but I don't want the SDK to suddenly cause a dozen pop up boxes to relogin.

Reply:
I'm also not familiar with SDK implementation practices in this space, however that sounds like a responsibility of the registered `SsoVerificationCallback` to perform any debouncing if necessary. I would imagine that the calls made to the sso/oidc services would also return throttling based errors in extreme cases.

Reply:
Lets check with other SDK implementations.

Reply:
This appears to be handled/guarded against in the [parent class (RefreshingAWSCredentials::GetCredentials)](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Credentials/RefreshingAWSCredentials.cs#L121), which wraps the call in a semaphore (prevents concurrent queries), and checks whether the first set of credentials has expired.

Reply:
This will work for a single instance of SSOAWSCredentials. If multiple SSOAWSCredentials are created for the same SSO configuration/profile you could get multiple logins. 

I'm fine with documenting for now best practices are to create a single instance. We might need to revisit this as we get customer feedback.